### PR TITLE
Switch to makefile-based doc build.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

My own attempt at switching to `make html` for docs build.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: doc/devel/documenting_mpl.rst [hunk 1]
```diff
@@ -16,18 +16,19 @@ are needed to build the documentation. They are listed in the file
...
+* numpydoc>=0.4
+* Pillow
+* sphinx-gallery>=0.1.12
+* graphviz
 
 .. note::
 
   * You'll need a minimal working LaTeX distribution for many examples to run.
-  * `Graphviz <http://www.graphviz.org/Download.php>`_ is not a python package, and needs
+  * `Graphviz <http://www.graphviz.org/Download.php>`_ is not a python package, and needs
```
Comment (QuLogic) [inline doc/devel/documenting_mpl.rst]: Unnecessary change?

File: doc/README.txt [hunk 2]
```diff
@@ -1,18 +1,20 @@
+.. highlight:: sh
+
 Matplotlib documentation
 ========================
 
-
```
Comment (jklymak) [inline doc/README.txt]: This seems to duplicate documenting_mpl.rst  Can all of this just be removed and replaced by reference?

File: doc/devel/release_guide.rst [hunk 3]
```diff
@@ -54,7 +54,7 @@ temporarily comment out the include and toctree glob; re-instate these after a
 release. Finally, make sure that the docs build cleanly ::
 
   pushd doc
-  python make.py html latex -n 16
+  make html latexpdf -n$(nproc)
```
Comment (QuLogic) [inline doc/devel/release_guide.rst]: `O=-n$(nproc)`? Currently, this will build `html` and `latexpdf` in parallel, but I'm not sure Sphinx will like two runs occurring in parallel.

File: doc/make.bat [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	echo.If you don't have Sphinx installed, grab it from
+	echo.http://sphinx-doc.org/
+	exit /b 1
+)
+
+%SPHINXBUILD% -M %1 %SOURCEDIR% %BUILDDIR% %SPHINXOPTS%
+goto end
+
+:help
+%SPHINXBUILD% -M help %SOURCEDIR% %BUILDDIR% %SPHINXOPTS%
```
Comment (QuLogic) [inline doc/make.bat]: Don't you need to check for sphinx build before running this too?

File: doc/devel/release_guide.rst (hunk 3 above)
Comment (jklymak) [reply doc/devel/release_guide.rst]: Ooops, I didn't check this one.   Good eye: that doesn't run at all on my machine.

Glad someone more knowledgeable is reviewing!

File: doc/devel/documenting_mpl.rst (hunk 1 above)
Comment (anntzer) [reply doc/devel/documenting_mpl.rst]: fixed, dunno how the spare space got there

File: doc/README.txt (hunk 2 above)
Comment (anntzer) [reply doc/README.txt]: yup, fixed

File: doc/devel/release_guide.rst (hunk 3 above)
Comment (anntzer) [reply doc/devel/release_guide.rst]: Good catch.

File: doc/make.bat (hunk 4 above)
Comment (anntzer) [reply doc/make.bat]: this should be reported to upstream but I fixed it locally.

**Comment by anntzer** [comment]
Comments addressed in force-pushed commit.  Thanks for the review.

File: doc/devel/release_guide.rst (hunk 3 above)
Comment (QuLogic) [reply doc/devel/release_guide.rst]: I think you only fixed the one below that didn't have this comment. 😛

Comment (anntzer) [reply doc/devel/release_guide.rst]: try again...

**Comment by anntzer** [comment]
rebased after the merge of #9311

File: doc/Makefile [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# Minimal makefile for Sphinx documentation
+#
+
+# You can set these variables from the command line.
+SPHINXOPTS    =
```
Comment (tacaswell) [inline doc/Makefile]: Can we default to `-W`?

Comment (anntzer) [reply doc/Makefile]: I would much prefer not: when building locally I can easily check from the log whether the build failed, but I'd rather not have the build (which can be quite lengthy) fail halfway, fix one error, restart a build, wait another 5min, fix the next error, etc.  (Well *I* know --allowsphinxwarnings, or, on the new system, how to override O=; but I think it's not the most new-contributor friendly way to do it...)
(xref https://github.com/sphinx-doc/sphinx/issues/3627)

**Comment by anntzer** [comment]
Minor edit to use `make -Cdoc ...` instead on pushd/popd in the release guide.

File: doc/Makefile (hunk 5 above)
Comment (tacaswell) [reply doc/Makefile]: The CI runs with it so new contributors are going to have to deal with the failures one way or the other and getting them locally by default seems nicer than having to go through CI

Comment (anntzer) [reply doc/Makefile]: Warnings are perfectly well printed without -W as well, the only effects of -W are that 1) the build stops after the first failure (not so nice for new contribs) and 2) the return status is nonzero in that case (relevant for CI, not so much for new contribs).

**Comment by QuLogic** [comment]
Seems to need a rebase.

**Comment by anntzer** [comment]
fixed

**Comment by tacaswell** [comment]
I prefer to default to fail-on-warning, but will not block the PR over it.

File: doc/Makefile (hunk 5 above)
Comment (jklymak) [reply doc/Makefile]: I can see both sides, but I guess I think it'd be more confusing to have it build locally and then fail on CI.  OTOH it'd be great to have the warning switch very clearly documented so knowledgable users can run that way and then clean all their errors in one compile.  

Besides, the current behaviour is to fail on warning by default.

**Comment by anntzer** [comment]
changed to -W by default

File: doc/make.py [hunk 6]
```diff
@@ -1,228 +0,0 @@
-#!/usr/bin/env python
```
Comment (jklymak) [inline doc/make.py]: OK, this all looks great to me, and it works, and the tests pass, so I think its good to go.  *But* do we need to remove `make.py`?  Someone may have a fancy script that uses it for something and want to keep it around, maybe with a big warning at the beginning that its going away soon?  

The converse of that argument is that it might be confusing to have both.  

I don't really care, but was just trying to think of what could go wrong.

Comment (anntzer) [reply doc/make.py]: I don't think we ever made any guarantees as to backcompatibility of tooling (and wouldn't want to do so either).

Comment (tacaswell) [reply doc/make.py]: Even I think we don't need to worry about this too much.  The downstream packagers (debian, fedora, etc) will need to adapt their build scripts, but it is clearly documented what to change it too.

**Comment by tacaswell** [comment]
Thanks @anntzer

**Comment by jklymak** [comment]
Yay!
